doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -126,114 +152,156 @@ Type declarations
 
-   This command defines a type with a manifest. There are four possible
-   kinds of such definitions: alias, variant, record and open variant types.
+     .. todo PR above: does it add an additional constructor for an existing o v type or
+        does it create both the type (and apparently the constructor?
 
-   .. productionlist:: coq
-      ltac2_typedef    : `ltac2_type`
-                       : [ `ltac2_constructordef` | ... | `ltac2_constructordef` ]
-                       : { `ltac2_fielddef` ; ... ; `ltac2_fielddef` }
-                       : [ .. ]
-      ltac2_constructordef   : `uident` [ ( `ltac2_type` , ... , `ltac2_type` ) ]
-      ltac2_fielddef         : [ mutable ] `ident` : `ltac2_type`
+   Without :n:`{| := | ::= }`
+     Defines an abstract type for use representing data from OCaml.  Not for
+     end users.
 
-   Aliases are just a name for a given type expression and are transparently
-   unfoldable to it. They cannot be recursive. The non-terminal
-   :production:`uident` designates identifiers starting with an uppercase.
+   :n:`with @tac2typ_def`
+     Permits definition of mutually recursive type definitions.
+
+   Each production of :token:`tac2typ_knd` defines one of four possible kinds
+   of definitions, respectively: alias, variant, open variant and record types.
+
+   Aliases are names for a given type expression and are transparently
+   unfoldable to it. They cannot be recursive.
+
+   .. The non-terminal :token:`uident` designates identifiers starting with an uppercase.
 
    Variants are sum types defined by constructors and eliminated by
    pattern-matching. They can be recursive, but the `rec` flag must be
    explicitly set. Pattern matching must be exhaustive.
 
+   Open variants can be extended with additional constructors using the `::=` form.
+
    Records are product types with named fields and eliminated by projection.
    Likewise they can be recursive if the `rec` flag is set.
 
-   .. cmdv:: Ltac2 Type {? @ltac2_typeparams } @ltac2_qualid ::= [ @ltac2_constructordef ]
+.. cmd:: Ltac2 @ external @ident : @ltac2_type := @string @string
+   :name: Ltac2 external
 
-      Open variants are a special kind of variant types whose constructors are not
-      statically defined, but can instead be extended dynamically. A typical example
-      is the standard `exn` type. Pattern matching on open variants must always include a catch-all
-      clause. They can be extended with this command.
+   Declares abstract terms.  Frequently, these declare OCaml functions
+   defined in |Coq| and give their type information.  They can also declare
+   data structures from OCaml.  It has no use for the end user.
+
+APIs
+~~~~
+
+Ltac2 provides over 150 API functions that provide various capabilities.  These
+are declared with :cmd:`Ltac2 external` in :n:`lib/coq/user-contrib/Ltac2/*.v`.
+For example, `Message.print` defined in `Message.v` is used to print messages:
```

Comment:
Why not
```suggestion
For example, :g:`Message.print` defined in `Message.v` is used to print messages:
```
to be a bit more semantically correct? (`g` for Gallina-like expressions, right?)

Also, is there any way to link `Message.v` to the stdlib doc for that file?  (And maybe do the same with `lib/coq/user-contrib/Ltac2/*.v` above)  cc @Zimmi48 @cpitclaudel ?

Reply:
Unfortunately not yet. And I'm not sure we're gonna keep `:g:` in the future given that the semantic information is not used. So I don't personally use it.